---
title: "GTEx_cis_Lasso_comparison"
author: "Natasha Santhanam"
date: "11/2/2021"
output: html_document
---

```{r}
suppressMessages(library(dplyr))
suppressMessages(library(glmnet))
suppressMessages((library(reshape2)))
suppressMessages(library(methods))
suppressMessages(library(RSQLite))
"%&%" <- function(a,b) paste(a,b, sep = "")

dir <- "/gpfs/data/im-lab/nas40t2/Data/GTEx/V8/GTEx_Analysis_v8_eQTL_expression_matrices/"
source("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/run_cis_Lasso.R")
```

#Make Phenotype file for GTex and subsample to number in Rats
```{r create separate txt files for each rxn}
set.seed(80)
pheno <- fread(dir %&% "Brain_Frontal_Cortex_BA9.v8.normalized_expression.bed.gz")
#subsample pheno
ids <- sample(colnames(pheno)[5:ncol(pheno)], 80)
pheno <- pheno %>% dplyr::select(c(gene_id, ids))

#pivot to right format where people are rows
long_pheno <- pheno %>% pivot_longer(!gene_id, names_to = "ids", values_to = "expr") %>% pivot_wider(names_from = gene_id, values_from = expr)
names <- long_pheno$ids
colnames(long_pheno)[2:ncol(long_pheno)] <- sapply(strsplit(colnames(long_pheno)[2:ncol(long_pheno)], "\\."), `[`, 1)

#create chunks of gene ids for cis lasso to use
long_pheno <- long_pheno %>% select(-c(ids))

genes <- colnames(long_pheno)
df <- data.frame(genes)
n = 50 #number of groups
split <- df %>% group_by(row_number() %/% n) %>% group_map(~ .x)

for(i in 1:length(split)) {
  tempo <- data.frame(split[i])
  write_tsv(tempo, "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_phenos/" %&% "phenolist"  %&% i %&% ".txt", col_names = FALSE)
}

#scale the phenotype data and save
scale_pheno <-  scale(long_pheno, center = TRUE, scale = TRUE) %>% as.data.frame() %>% mutate(IID = names, .before = "ENSG00000227232")
#write_tsv(scale_pheno, "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_Brain_Frontal_Cortex_BA9.v8.scaled_expression.txt", col_names = TRUE)
```


#Convert genotype file to Plink 2 pgen format
Make sure to subsample geno file too
```{bash, eval=FALSE}
plink2 --bfile GTEx_maf_0.01 --keep subsample_GTEx_cortex_ids.txt --make-pgen vzs --out GTEx_maf_0.01_pgen
```

# Generate SNP annotation and Gene annotation files

```{r create snp annot, eval=FALSE}
#can create snp annotation from bim file 
bim <- fread("/gpfs/data/im-lab/nas40t2/Data/GTEx/V8/genotype/plink_files/GTEx_maf_0.01.bim")
colnames(bim) <- c("chr", "varID", "CM", "pos", "refAllele", "effectAllele")
#saveRDS(bim, "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_snp_annotation.RDS")

# gene annotation file
gene_annot <- read_tsv("/gpfs/data/im-lab/nas40t2/natasha/GTEX_Analysis/annotations_gencode_v26.tsv", col_names = TRUE)
gene_annot <- gene_annot %>% rename(chr = chromosome)
#saveRDS(gene_annot, "/gpfs/data/im-lab/nas40t2/natasha/GTEX_Analysis/annotations_gencode_v26.RDS")
```


#Run cis Lasso 

```{bash run cis Lasso, eval=FALSE}
#!/bin/bash

#PBS -N cis_Lasso_GTEx
#PBS -S /bin/bash
#PBS -l walltime=48:00:00
#PBS -l mem=64gb
#PBS -l nodes=1:ppn=4

# SPECIFY LOGGING BEHAVIOR

#PBS -o /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/logs/cis_Lasso_GTEx_${PBS_JOBID}.out
#PBS -e /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/logs/cis_Lasso_GTEx_${PBS_JOBID}.err

module load gcc/6.2.0 
module load python/3.8.1
#module load miniconda3/4.7.10

source ~/.bashrc
conda activate /gpfs/data/im-lab/nas40t2/festus/my_envs/snpnet

echo "`date` Starting"

Rscript /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/run_cis_Lasso.R \
--phenotype_table /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_Brain_Frontal_Cortex_BA9.v8.scaled_expression.txt \
--genotype /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_plink2/GTEx_maf_0.01_pgen \
--pheno_list /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_phenos/phenolist${chk}.txt \
--snpnet_config /gpfs/data/im-lab/nas40t2/festus/metabolomics/guardian/MetaboXcan/scripts/gw_lasso/snpnet_config.yaml \
--gene_annot /gpfs/data/im-lab/nas40t2/natasha/GTEX_Analysis/annotations_gencode_v26.RDS \
--snp_annot /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_snp_annotation.RDS \
--indiv_col IID \
--nfold 5 \
--inner_nfold 5 \
--mode cv_performance \
--output_prefix /scratch/nsanthanam1/cis_Lasso_GTEx/cis_GTEx_lasso_perf.${chk}

Rscript /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/run_cis_Lasso.R \
--phenotype_table /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_Brain_Frontal_Cortex_BA9.v8.scaled_expression.txt \
--genotype /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_plink2/GTEx_maf_0.01_pgen \
--pheno_list /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_phenos/phenolist${chk}.txt \
--snpnet_config /gpfs/data/im-lab/nas40t2/festus/metabolomics/guardian/MetaboXcan/scripts/gw_lasso/snpnet_config.yaml \
--gene_annot /gpfs/data/im-lab/nas40t2/natasha/GTEX_Analysis/annotations_gencode_v26.RDS \
--snp_annot /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/cis_Lasso/GTEx_snp_annotation.RDS \
--indiv_col IID \
--nfold 5 \
--inner_nfold 5 \
--mode model_training \
--output_prefix /scratch/nsanthanam1/cis_Lasso_GTEx/cis_GTEx_lasso_weights.${chk}
```


```{bash submit jobs for all phenos, eval=FALSE}
for i in {1..495}; do qsub -v chk=${i} cis_Lasso_GTEx.pbs ; done

for i in {2..200}; do qsub -v chk=${i} cis_Lasso_GTEx.pbs ; done
```
