---
title: "GW_Ridge_Rats"
author: "Natasha Santhanam"
date: "9/10/2021"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(data.table)
library(readxl)
library(qqman)
library(glmnet)
library(data.table)
"%&%" = function(a,b) paste(a,b,sep="")
#source("/Users/natashasanthanam/Github/rat-genomic-analysis/rat_heritability_analysis.Rmd")
```
To Run GW Ridge - you need these inputs
genotype - GRM for all 80 Rats (include grm.bin, grm.N.bin, grm.id)
         - BED file 
phenotype - parquet file with phenotype that is also normalized + covariates regressed out
  rows are people, columns is expression with a column for IID

```{r gather input files, eval=FALSE}
geno_files <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/plink_files/"
pheno_files <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/expression/"
grm_dir <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/GEMMA/Ac/GRMs/"
```

```{bash create GRM for Ac Tissue in rats, eval = FALSE}
plink --bfile $geno_files/rat_genome_Ac  --make-grm-gz --out  --out grm_dir/rat_Palmer_Ac_grm
```

```{r create phenotype parquet file, eval = FALSE}
expressionAc <- readRDS("/Users/natashasanthanam/Github/rat-genomic-analysis/data/Ac_expression_transformed.RDS")

#do this in cis

#center and scale expression data
expressionAc <- scale(expressionAc, center = TRUE, scale = TRUE)
expressionAc <- as.data.frame(expressionAc)
expressionAc <- expressionAc %>% mutate(IID = rownames(expressionAc), .before = "ENSRNOG00000000007")
write_parquet(expressionAc, "/Users/natashasanthanam/Github/rat-genomic-analysis/data/rat_Palmer_Ac_normalized.parquet")
```

```{bash GW Run Ridge, eval=FALSE}
#!/bin/bash

#PBS -N Palmer-Ac-scale_ridge
#PBS -S /bin/bash
#PBS -l walltime=96:00:00
#PBS -l mem=64gb
#PBS -l nodes=1:ppn=1

# SPECIFY LOGGING BEHAVIOR

#PBS -o /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/GW_Ridge/logs/${PBS_JOBNAME}.${PBS_JOBID}.log
#PBS -e /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/GW_Ridge/logs/${PBS_JOBNAME}.${PBS_JOBID}.err

module load gcc/6.2.0 
module load miniconda3/4.7.10


source ~/.bashrc
source ~/.bash_profile
source activate /gpfs/data/im-lab/nas40t2/bin/envs/tensorqtl/


echo "`date` Starting"
python3 /gpfs/data/im-lab/nas40t2/festus/metabolomics/guardian/MetaboXcan/scripts/gw_ridge/run_gw_ridge.py \
--gcta_grm_prefix /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/GEMMA/Ac/GRMs/rat_Palmer_Ac_grm \
--phenotype_parquet /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/expression/rat_Palmer_Ac_normalized.parquet \
--geno_bed_pattern /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/plink_files/rat_genome_Ac.bed \
--nfold 3 3 \
--output /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/GW_Ridge/ridge_cv_perf-Palmer_Ac_rat-scale.tsv.gz

echo "Training model"


python3 /gpfs/data/im-lab/nas40t2/festus/metabolomics/guardian/MetaboXcan/scripts/gw_ridge/run_gw_ridge.py \
--gcta_grm_prefix /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/GEMMA/Ac/GRMs/rat_Palmer_Ac_grm \
--phenotype_parquet /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/expression/rat_Palmer_Ac_normalized.parquet \
--geno_bed_pattern /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/plink_files/rat_genome_Ac.bed \
--train_full_model \
--nfold 3 3 \
--output /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/GW_Ridge/ridge_model_weights-Palmer_Ac_rat-scale.tsv.gz
```

Run Ridge Regression on only cis-SNPs
```{r read in files, eval=FALSE}
expression_RDS <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/expression/Ac_expression_transformed.RDS"
geno_file <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/genotype_files/genotype.txt"
gene_annot_RDS <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/gtf.RDS"
snp_annot_RDS <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/phyMap.RDS"

n_k_folds <- 10
alpha <- 0
snpset <- "1KG_snps"
window <- 1000
tis <- "Ac"
out_dir <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/GW_Ridge/"
```


```{r run only cis-Ridge, eval=FALSE}
library(tidyverse)
library(data.table)
library(glmnet)
"%&%" = function(a,b) paste(a,b,sep="")

expression_RDS <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/expression/Ac_expression_transformed.RDS"
geno_file <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/genotype_files/genotype.txt"
gene_annot_RDS <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/gtf.RDS"
snp_annot_RDS <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/phyMap.RDS"

n_k_folds <- 10
alpha <- 0
snpset <- "1KG_snps"
window <- 1000
tis <- "Ac"
out_dir <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/GW_Ridge/perf/"


expression <- readRDS(expression_RDS)
class(expression) <- 'numeric'
genotype <- read.table(geno_file, header = TRUE, stringsAsFactors = FALSE)
# Transpose genotype for glmnet
names(genotype) <- substring(names(genotype), 2)
genotype <- genotype[match(rownames(expression), colnames(genotype))]
genotype <- t(genotype)
gene_annot <- readRDS(gene_annot_RDS)
#gene_annot <- subset(gene_annot, gene_annot$chr == chrom)
snp_annot <- readRDS(snp_annot_RDS)

rownames(gene_annot) <- gene_annot$Gene
# Subset expression data to only include genes with gene_info
expression <- expression[, intersect(colnames(expression), rownames(gene_annot))]
exp_samples <- rownames(expression)
exp_genes <- colnames(expression)
n_samples <- length(exp_samples)
n_genes <- length(exp_genes)
groupid <- sample(1:n_k_folds, length(exp_samples), replace = TRUE)

resultsarray <- array(0, c(length(exp_genes), 9))
dimnames(resultsarray)[[1]] <- exp_genes
resultscol <- c("gene", "alpha", "cvm", "lambda.iteration", "lambda.min", "n.snps", "R2", "pval", "cor")
dimnames(resultsarray)[[2]] <- resultscol
workingbest <- out_dir %&% "working_TW_" %&% tis %&% "_exp_" %&% n_k_folds %&% "-foldCV_elasticNet_alpha" %&% alpha %&% "_" %&% snpset  %&% ".txt"
write(resultscol, file = workingbest, ncolumns = 9, sep = "\t")

weightcol <- c("gene","rsid","ref","alt","beta","alpha")
workingweight <- out_dir %&% "TW_" %&% tis %&% "_elasticNet_alpha" %&% alpha %&% "_" %&% snpset %&% "_weights" %&% ".txt"
write(weightcol, file = workingweight, ncol = 6, sep = "\t")

for (i in 1:length(exp_genes)) {
  cat(i, "/", length(exp_genes), "\n")
  gene <- exp_genes[i]
  # Reduce genotype data to only include SNPs within specified window of gene.
  geneinfo <- gene_annot[gene,]
  start <- geneinfo$Start - window
  end <- geneinfo$End + window
  # Pull cis-SNP info
  cissnps <- subset(snp_annot, snp_annot$pos >= start & snp_annot$pos <= end)
  # Pull cis-SNP genotypes
  cisgenos <- genotype[,intersect(colnames(genotype), cissnps$varID), drop = FALSE]
  # Reduce cisgenos to only include SNPs with at least 1 minor allele in dataset
  cm <- colMeans(cisgenos, na.rm = TRUE)
  minorsnps <- subset(colMeans(cisgenos), cm > 0 & cm < 2)
  minorsnps <- names(minorsnps)
  cisgenos <- cisgenos[,minorsnps, drop = FALSE]
  if (ncol(cisgenos) < 2) {
    # Need 2 or more cis-snps for glmnet.
    bestbetas <- data.frame()
  } else {
    # Pull expression data for gene
    exppheno <- expression[,gene]
    # Scale for fastLmPure to work properly
    exppheno <- scale(exppheno, center = TRUE, scale = TRUE)
    exppheno[is.na(exppheno)] <- 0
    rownames(exppheno) <- rownames(expression)
    # Run Cross-Validation
    # parallel = TRUE is slower on tarbell
    bestbetas <- tryCatch(
      { fit <- cv.glmnet(as.matrix(cisgenos),
                         as.vector(exppheno),
                         nfolds = n_k_folds,
                         alpha = alpha,
                         keep = TRUE,
                         foldid = groupid,
                         parallel = FALSE)
      # Pull info from fit to find the best lambda
      fit.df <- data.frame(fit$cvm, fit$lambda, 1:length(fit$cvm))
      # Needs to be min or max depending on cv measure (MSE min, AUC max, ...)
      best.lam <- fit.df[which.min(fit.df[,1]),]
      cvm.best <- best.lam[,1]
      lambda.best <- best.lam[,2]
      # Position of best lambda in cv.glmnet output
      nrow.best <- best.lam[,3]
      # Get the betas from the best lambda value
      ret <- as.data.frame(fit$glmnet.fit$beta[,nrow.best])
      ret[ret == 0.0] <- NA
      # Pull the non-zero betas from model
      as.vector(ret[which(!is.na(ret)),])
      },
      error = function(cond) {
        # Should fire only when all predictors have 0 variance.
        message('Error with gene ' %&% gene %&% ', index ' %&% i)
        message(geterrmessage())
        return(data.frame())
      }
    )
  }
  if (length(bestbetas) > 0) {
    names(bestbetas) <- rownames(ret)[which(!is.na(ret))]
    # Pull out the predictions at the best lambda value.
    pred.mat <- fit$fit.preval[,nrow.best]
    res <- summary(lm(exppheno~pred.mat))
    rsq <- res$r.squared
    pval <- res$coef[2,4]
    cor <- cor(exppheno, pred.mat)
    resultsarray[gene,] <- c(gene, alpha, cvm.best, nrow.best, lambda.best, length(bestbetas), rsq, pval, cor)
    # Output best shrunken betas for PrediXcan
    bestbetalist <- names(bestbetas)
    bestbetainfo <- snp_annot[bestbetalist,]
    betatable <- as.matrix(cbind(bestbetainfo,bestbetas))
    # write_covariance(gene, cisgenos, betatable[,"rsid"], betatable[,"varID"], covariance_out)
    # Output "gene", "rsid", "refAllele", "effectAllele", "beta"
    # For future: To change rsid to the chr_pos_ref_alt_build label, change "rsid" below to "varID".
    betafile <- cbind(gene,betatable[,"rsid"],betatable[,"refAllele"],betatable[,"effectAllele"],betatable[,"bestbetas"], alpha)
    # Transposing betafile necessary for correct output from write() function
    write(t(betafile), file = workingweight, ncolumns = 6, append = TRUE, sep = "\t")
    write(resultsarray[gene,], file = workingbest, ncolumns = 9, append = TRUE, sep = "\t")
  } else {
    resultsarray[gene,1] <- gene
    resultsarray[gene,2:9] <- c(alpha,NA,NA,NA,0,NA,NA,NA)
  }
}
}
```




 
