---
title: "mice_eQTL_anaylsis"
author: "Natasha Santhanam"
date: "4/20/2021"
output: html_document
---

Run eQTL manually with mice AIL genotype data and str expression data and check with what was generated by Palmer Lab

```{r setup, include=FALSE}
library(tidyverse)
library(vroom)
library(MatrixEQTL)
library(dplyr)
library(RSQLite)
library(ggpubr)

setwd("/Users/natashasanthanam/Github/rat-genomic-analysis/data/")
box_dir <- "/Users/natashasanthanam/Box/imlab-data/Projects/PTRS-PGRS-Rosetta/Data-From-Abe-Palmer-Lab/G50-56_LGxSM_AIL_GWAS/"
"%&%" = function(a,b) paste(a,b,sep="")

load("/Users/natashasanthanam/Downloads/eigenMT.results.eqtl.RData")
```

Read in genotype and expression data
```{r}
dosages <- read_tsv(box_dir %&% "ail.genos.dosage.gwasSNPs.txt", col_names = FALSE)
pheno <- read_tsv(box_dir %&% "ail.phenos.final.txt", col_names = TRUE)
expr <- read.table(box_dir %&% "strNormCounts.expr.txt")
str_abrv <- str %>% filter(pvalue < 1e-6)


colnames(dosages)= c("chr", "pos", "ref", "alt", pheno$id)
gt <- dosages[,c(intersect(colnames(dosages)[5:ncol(dosages)], rownames(expr)))]

#check alignment
table(rownames(expr) == colnames(gt))

dosages$id = paste(dosages$chr, dosages$pos, sep = ".")
gt <- cbind(dosages$chr, dosages$pos, dosages$id, dosages$ref, dosages$alt, gt)
rownames(gt) = gt$`dosages$id`
```

Inverse normalize expression data
```{r}
invnorm = function(x) {
  if(is.null(dim(x))) res = invnorm.vector(x) else
  res=apply(x,2,invnorm.vector)
  res
}
invnorm.vector = function(x) {yy = rank(x)/(length(x)+1); qnorm(yy)}

expr_norm = invnorm(expr)
```



Boxplots of expression levels of gene against snp for normalized expresion and non-normalized
```{r}
#make abreviated genotype file to match the top eqtl snps
gt <- as.data.frame(gt)
gt <- gt[,-c(1:5)]

snps = c("chr1.117459320", "chr10.128477803", "chr12.106130002") 
genes = c("Clasp1", "Rps26", "Vrk1")
par(mfrow=c(1,length(snps)))

df <- as.data.frame(matrix(NA, 169, 2)) 
df_norm <- as.data.frame(matrix(NA, 169,2))
plot_list = list()
plot_list_norm = list()
my_comparisons <- list( c("0", "1"), c("1", "2"), c("0", "2") )

for (index in seq(length(snps))){
genotype = gt[snps[index],]
genotype = t(genotype)
expression = expr[,genes[index]]
expression_norm = expr_norm[,genes[index]]
df[,1] = round(genotype)
df[,2] = expression
colnames(df) = c("X1", "X2")
df_norm[,1] = round(genotype)
df_norm[,2] = expression_norm
colnames(df_norm) = c("X1", "X2")

p = ggboxplot(df, x = "X1", y = "X2", color = "X1", palette = "jco", add = "jitter") + 
  ggtitle(paste( snps[index], "vs", genes[index], sep=" ") ) +
  xlab("genotype") + ylab("expression") 
 plot_list[[index]] = p + stat_compare_means(comparisons = my_comparisons, method = "t.test")
 
p_norm = ggboxplot(df_norm, x = "X1", y = "X2", color = "X1", palette = "jco", add = "jitter") + 
  ggtitle(paste( "Normalized", snps[index], "vs", genes[index], sep=" ") ) +
  xlab("genotype") + ylab("expression") 
 plot_list_norm[[index]] = p_norm + stat_compare_means(comparisons = my_comparisons, method = "t.test")
}


ggarrange(plot_list[[1]],plot_list[[2]], plot_list[[3]], plot_list_norm[[1]],plot_list_norm[[2]], plot_list_norm[[3]], ncol = 3, nrow = 2)


```

```{r}
snps <- str_abrv$rs
genes = str_abrv$gene_name
output <- matrix(ncol=3, nrow=length(snps))
output_norm <- matrix(ncol=3, nrow=length(snps))

for(i in 1:length(snps)){
  if(genes[i] %in% colnames(expr)) {
  genotype = gt[snps[i],]
  output[i, 1] <-  snps[i]
  output[i, 2] <-  genes[i]
  output_norm[i, 1] <-  snps[i]
  output_norm[i, 2] <-  genes[i]
  genotype = t(genotype)
  lm_gene = lm(expr[,genes[i]] ~ genotype)
  lm_norm = lm(expr_norm[,genes[i]] ~ genotype)
  output[i, 3] <-  summary(lm_gene)$coefficients[2,4]
  output_norm[i, 3] <-  summary(lm_norm)$coefficients[2,4]
  }
  else {
    i = i+1
  }
}
```

cis-Eqtl

```{r}
#Using the MatrixEQTL functions import gene expression and genotype data as “SlicedData” objects. Verify that the samples are ordered correctly for genotype and phenotype

snp_values = gt[,6:ncol(gt)]
gene_values = t(expr)

snps <- SlicedData$new()
snps$CreateFromMatrix(as.matrix(snp_values))
gene <- SlicedData$new()
gene$CreateFromMatrix(as.matrix(gene_values))

all(colnames(snps) == colnames(gene))
```


```{r}
# load snp and gene annotation files
snps$ResliceCombined(1000)
gene$ResliceCombined(1000)
probePos <- read_tsv(data_dir %&% "mice_gtf" ) %>% select(c(gene_name, seqnames, start, end))
snpPos <- read_tsv(box_dir %&% "AIL.gwasSNPs.rsids.txt", col_names = FALSE)
snpPos <- snpPos %>% select(c(X2, X1, X3))


meq <- Matrix_eQTL_main(snps = snps, gene=gene, output_file_name=NULL,
output_file_name.cis=NULL, useModel = modelLINEAR, verbose = TRUE, pvOutputThreshold = 1e-6, 
pvOutputThreshold.cis=1e-3, snpspos=as.data.frame(snpPos), genepos=as.data.frame(probePos), cisDist = 1e6, pvalue.hist = "qqplot", min.pv.by.genesnp = FALSE, noFDRsaveMemory = FALSE)


# trans eqtl
eQTL_trans = Matrix_eQTL_engine(
snps = snps,
gene = gene,
output_file_name = output_file_name,
pvOutputThreshold = 1e-8,
useModel = modelLINEAR,
verbose = TRUE,
pvalue.hist = 100)
```

70,663 eQTLs for DBOB data with 21,781 genes and 289 rats and 137,406 snps with a pval < 1e-20
5344 cis qtls for Palmer data with ensembl gene annotation and inverse normalzied expressiond data with pval < 1e-6

Generate extra and weight files from prediction models generated by Dan Cox
```{r}
#read in snp annotation file
snpPos <- read_tsv(box_dir %&% "AIL.gwasSNPs.rsids.txt", col_names = FALSE)
colnames(snpPos) = c("chr", "variant", "pos", "alt", "rsid")

#get all weights file for every gene
ge.dir <- "/scratch/nsanthanam1/exp_pred_model"
files <- list.files(path = ge.dir, pattern = ".txt", full.names = TRUE)

#match variant ids from snp annotation file to ids in models
snpPos$variant = paste(sapply(strsplit(snpPos$variant, "\\."), `[`, 1), sapply(strsplit(snpPos$variant, "\\."), `[`, 2), sep = ":")

#bind all the models together with variants from snpid file
df <- data.frame()
for(fila in files) {
  gene <- vroom(fila, col_names = TRUE) %>% mutate(gene = sapply(strsplit(sapply(strsplit(fila, "\\."), `[`, 1), "/"), `[`, 5))
  check <- inner_join( gene, snpPos, by = "variant") 
  df <- rbind(check, df)
}

#generate extra file with gene name and number of snps
tempo <- data.frame(gene = as.character(), n.snps.model = numeric(), pred.perf.R2 = numeric(), pred.perf.pval = numeric() )
for(i in 1:length(files)) {
  df <- vroom(files[i], col_names = TRUE) 
   gene = sapply(strsplit(sapply(strsplit(files[i], "\\."), `[`, 1), "/"), `[`, 5)
   tempo[nrow(tempo) + 1,1] <- gene
   tempo[nrow(tempo),2] <-  nrow(df)
   tempo[nrow(tempo),3] <-  mean(df$cor_r)
   tempo[nrow(tempo),4] <-  mean(df$cor_p)
   
}
```


```{r}
#Add ensembl gene_names to each ensid from file 
ensembl <- useEnsembl(biomart = "genes")
ensembl <- useEnsembl(biomart = "ensembl", 
                   dataset = "mmusculus_gene_ensembl", 
                   mirror = "asia")

genes <-  as.data.frame(extra$ensembl_gene_id)

gene_IDs_mice <- getBM(attributes= c('external_gene_name', 'ensembl_gene_id', 'gene_biotype'), 
      filters = '', 
      values = genes, 
      mart = ensembl)

extra <- left_join(extra, gene_IDs_mice, by = "ensembl_gene_id")
extra <- extra %>% mutate(pred.perf.R2 = NA, pred.perf.pval = NA, pred.perf.qval = NA) %>% select(c("gene", "gene_name", "gene_type","n.snps.in.model")) 


#Use snp annotation file and add a column for alt allele from dosage file
ref <- dosages %>% select(c(X1, X2, X3))
check <- ref[match(weights$variant, rownames(ref))  ,]

weights <- cbind(check$X3, weights)
weights <- weights %>% select(gene, rsid, variant, ref, alt, weight )
#match variant ids to genotype file
weights$variant =   paste(sapply(strsplit(weights$variant, ":"), `[`, 1), sapply(strsplit(weights$variant, ":"), `[`, 2), sep = ".")
colnames(weights) = c("gene", "rsid", "varID", "ref_allele", "eff_allele", "weight")
t = 0
for(i in 1:length(files)) {
  df <- read_tsv(files[i], col_names = TRUE) 
  t <- t + nrow(df)
  }
```

422,212 snps in the model ~not the same snps 
523,028 snps in snp annotation file
9,555 snps are in common between models and annotation file

18022 genes from model
14364 in extra file (had to add ensembl ids)


Make dbs
```{bash, eval=FALSE}
sqlite3 mice_model.db "VACUUM;"
```

```{r}
weights <- read.table(data_dir %&% "weights", header = T, stringsAsFactors = F)
extra <- read.table(data_dir %&% "extra", header = T, stringsAsFactors = F)

driver <- dbDriver('SQLite')
#made a db table in for just extra table
conn <- dbConnect(drv = driver,  data_dir %&% 'mouse_model.db')
  dbWriteTable(conn, 'weights', weights, overwrite = TRUE)
  dbWriteTable(conn, 'extra', extra, overwrite = TRUE)
  dbDisconnect(conn)
```

```{bash, eval=FALSE}
conda activate /gpfs/data/im-lab/nas40t2/natasha/envs/predixcan
export PRE=/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/mice_predixcan
printf "Predict expression\n\n"

python3 /gpfs/data/im-lab/nas40t2/natasha/GTEX_Analysis/MetaXcan/software/Predict.py \
--model_db_path $PRE/mouse_model.db  \
--model_db_snp_key varID \
--vcf_genotypes $PRE/mice_geno.vcf  \
--vcf_mode genotyped \
--prediction_output $PRE/mouse__predict_new.txt \
--prediction_summary_output $PRE/mouse__summary_new.txt \
--verbosity 9 \
--throw
```

Read in predicted expression
```{r}
expr_pred <- read_tsv(data_dir %&% "mouse__predict_new.txt", col_names = TRUE)

genes <-  as.data.frame(colnames(expr_pred)[3:ncol(expr_pred)])
gene_names_mice <- getBM(attributes= c('external_gene_name', 'ensembl_gene_id'), 
      filters = '', 
      values = genes, 
      mart = ensembl)

gene_names_mice <- gene_names_mice[match(colnames(expr_pred)[3:ncol(expr_pred)], gene_names_mice$ensembl_gene_id) ,]
colnames(expr_pred)[3:ncol(expr_pred)] = gene_names_mice$external_gene_name

expr_pred$FID = sapply(strsplit(expr_pred$FID, "_"), `[`, 1)
expr_pred$IID = sapply(strsplit(expr_pred$IID, "_"), `[`, 1)
```

Perform Spearman Correlation 
```{r}
# match mice and genes between predicted expression and measured one
rownames(expr_pred) <- expr_pred$FID
expr_abrv <- expr_pred[match(rownames(expr), rownames(expr_pred)), ]
genes_in_common <- intersect(colnames(expr_abrv), colnames(expr))

expr_abrv <- as.matrix(sapply(expr_abrv[,3:5231], as.numeric))
rownames(expr_abrv) <- expr_pred$FID
expr_abrv <- expr_abrv[,match(genes_in_common, colnames(expr_abrv))]

expr <- expr[,match(genes_in_common, colnames(expr))]
#check to see gene names and ids are in the right order between two
table(colnames(expr) == colnames(expr_abrv))
table(rownames(expr) == rownames(expr_abrv))

expr <- t(expr)
expr_abrv <- t(expr_abrv)
cor.matrix <- cor(expr, expr_abrv,  method = "spearman")
```







