---
title: "Prediction_Pipeline_Mice"
author: "Natasha Santhanam"
date: "5/20/2021"
output: html_document
---

```{r}
library(tidyverse)
library(devtools)
library(broom)
library(data.table)
"%&%" = function(a,b) paste(a,b,sep="")
data.dir <- "/Users/natashasanthanam/Github/rat-genomic-analysis/data/"
```
#Done in CRI

```{r}
box_dir <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/DBOB_data/"
setwd("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/prediction_pipeline/DOB_pipeline/")
```

Gather Genotype, Gene epression data, snp annotation file and gtf (made with ensembl notation)

```{r}
geno <- read_csv(box_dir %&% "mice_DBOB_bimbam", col_names = FALSE)
gex <- read.table(box_dir %&% "Attie_232_Attie_DO_Islets__expr_norm.txt" )
mice_pheno <- read_tsv(box_dir %&% "mice_panc_pheno.txt", col_names = TRUE)
```

Format gene annotation
```{r}
gtf <- read_tsv("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/mice_pancreas/mice_pancreas_gtf.txt", col_names =TRUE)

gene_annotation <- gtf %>% select(c(chr, gene_id, gene_name, start, end))
rownames(gene_annotation) = gtf$gene_id
```

Format snp annotation
```{r}
phyMap <- read_csv(box_dir %&% "mice_DBOB_bimbam", col_names = FALSE) %>% select(c(X3, X1, X2, X4, X5))
colnames(phyMap) = c("snp", "chr", "pos", "refAllele", "effectAllele")  

phyMap <- within(phyMap,  varID <- paste(chr, pos, refAllele, effectAllele, sep="_"))
phyMap$rsid = phyMap$varID
phyMap <- phyMap[!duplicated(phyMap$varID),]
rownames(phyMap) = phyMap$varID

#remove snps that dont have ref or effect allele
phyMap <- phyMap %>% filter(refAllele != "-" | effectAllele != "-")
```

```{r}
# Splitting the snp annotation file by chromosome
s <- setNames(split(phyMap, phyMap$chr), paste0("snp_annot.chr", unique(phyMap$chr)))
list2env(s, globalenv())
```

```{r}
# writing the genotype file to a .txt file so that we can separate it by chromosome using our geneotype parse script.
colnames(geno) = c("chr", "pos", "marker", "ref", "eff", mice_pheno$FID)
rownames(geno) = geno$marker
geno <- geno[intersect(geno$marker, phyMap$snp), ] %>% select(c(marker, chr, pos, colnames(geno)[6:ncol(geno)]))
rownames(geno) = rownames(phyMap)
write.table(geno, file = "genotype.txt", sep = "\t", col.names = TRUE, row.names = TRUE)

gex <- gex[intersect(colnames(geno), rownames(gex)),]
```

```{bash}
#Splitting the genotype file by chromosome - run this from the rat_genomic_alaysis directory
python rat-genomic-analysis/scripts/split_genotype_by_chr.py DOB_pipeline/genotypes/genotype.txt DOB_pipeline/genotypes/genotype
```

```{r}
# Writing the gene expression files to csv files to be used for PEER Factor analysis
write.table(gex, file = 'gex.csv', sep = ",", col.names = TRUE, row.names = FALSE)
```

```{bash}
# Now we are performing PEER factor analysis on each tissue choosing 7 factors
/gpfs/data/im-lab/nas40t2/yanyul/softwares/peer/bin/peertool -f DOB_pipeline/gex.csv -n 7 -o /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/prediction_pipeline/DOB_pipeline/peer_mice --has_header
```

Read in Peer Factors
```{r}
peer_factors= read.csv(file = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/prediction_pipeline/DOB_pipeline/peer_mice/X.csv", header = FALSE)
```

```{r}
# Manipulating the PEER factor files so we can append to covariate file
rownames(peer_factors) = rownames(gex)
colnames(peer_factors) = c('PF1', 'PF2', 'PF3', 'PF4', 'PF5', 'PF6', 'PF7')
```

```{r}
#initializing matrices to be filled with t-stats, p_vals, and residuals of the regression of each gene vs. the covariates for each tissue. t-stat and p-val matrices are just for diagnostics
# t_stats = matrix(nrow = 13, ncol = length(colnames(gex)))
# p_vals = matrix(nrow = 13, ncol = length(colnames(gex)))
expression = gex
```

```{r}
# Regressing out the covariates and saving the residuals as the new expression for each tissue

for (i in 1:length(colnames(gex))) {
    fit = lm(gex[,i] ~  peer_factors$PF1 + peer_factors$PF2 + peer_factors$PF3 + peer_factors$PF4 + peer_factors$PF5 + peer_factors$PF6 + peer_factors$PF7)
    expression[,i] <- fit$residuals
}
```

```{r}
# Saving the expression RDS objects to be used as arguments in the script
saveRDS(as.matrix(expression), "mice_expression_transformed.RDS")
```

```{r}
#set snp annotation rownames 
rownames(snp_annot.chr1) = snp_annot.chr1$varID
rownames(snp_annot.chr2) = snp_annot.chr2$varID
rownames(snp_annot.chr3) = snp_annot.chr3$varID
rownames(snp_annot.chr4) = snp_annot.chr4$varID
rownames(snp_annot.chr5) = snp_annot.chr5$varID
rownames(snp_annot.chr6) = snp_annot.chr6$varID
rownames(snp_annot.chr7) = snp_annot.chr7$varID
rownames(snp_annot.chr8) = snp_annot.chr8$varID
rownames(snp_annot.chr9) = snp_annot.chr9$varID
rownames(snp_annot.chr10) = snp_annot.chr10$varID
rownames(snp_annot.chr11) = snp_annot.chr11$varID
rownames(snp_annot.chr12) = snp_annot.chr12$varID
rownames(snp_annot.chr13) = snp_annot.chr13$varID
rownames(snp_annot.chr14) = snp_annot.chr14$varID
rownames(snp_annot.chr15) = snp_annot.chr15$varID
rownames(snp_annot.chr16) = snp_annot.chr16$varID
rownames(snp_annot.chr17) = snp_annot.chr17$varID
rownames(snp_annot.chr18) = snp_annot.chr18$varID
rownames(snp_annot.chr19) = snp_annot.chr19$varID
rownames(snp_annot.chr20) = snp_annot.chr20$varID
```


```{r}
# Saving the SNP annotation RDS objects to be used as arguments in the script - too lazy to write a for loop 
saveRDS(snp_annot.chr1, "snp_annot.chr1.RDS")
saveRDS(snp_annot.chr2, "snp_annot.chr2.RDS")
saveRDS(snp_annot.chr3, "snp_annot.chr3.RDS")
saveRDS(snp_annot.chr4, "snp_annot.chr4.RDS")
saveRDS(snp_annot.chr5, "snp_annot.chr5.RDS")
saveRDS(snp_annot.chr6, "snp_annot.chr6.RDS")
saveRDS(snp_annot.chr7, "snp_annot.chr7.RDS")
saveRDS(snp_annot.chr8, "snp_annot.chr8.RDS")
saveRDS(snp_annot.chr9, "snp_annot.chr9.RDS")
saveRDS(snp_annot.chr10, "snp_annot.chr10.RDS")
saveRDS(snp_annot.chr11, "snp_annot.chr11.RDS")
saveRDS(snp_annot.chr12, "snp_annot.chr12.RDS")
saveRDS(snp_annot.chr13, "snp_annot.chr13.RDS")
saveRDS(snp_annot.chr14, "snp_annot.chr14.RDS")
saveRDS(snp_annot.chr15, "snp_annot.chr15.RDS")
saveRDS(snp_annot.chr16, "snp_annot.chr16.RDS")
saveRDS(snp_annot.chr17, "snp_annot.chr17.RDS")
saveRDS(snp_annot.chr18, "snp_annot.chr18.RDS")
saveRDS(snp_annot.chr19, "snp_annot.chr19.RDS")
saveRDS(snp_annot.chr20, "snp_annot.chr20.RDS")
```

```{r}
# Saving the gene annotation RDS object to be used as an argument in the script
saveRDS(gene_annotation, "gene_annotation.RDS")
```


```{bash}
# Creating the meta data file for each tissue 
python rat-genomic-analysis/scripts/create_meta_data.py --geno "DOB_pipeline/genotypes/genotype.txt" --expr "DOB_pipeline/mice_expression_transformed.RDS" --snpset "1KG" --rsid_label 1 --window 1000000 --out_prefix "DOB_pipeline/Results/allMetaData/mice_DO" 
```

```{bash}
# Running the model training script for each tissue/chromosome pair
for i in {1..20}
do
  Rscript rat-genomic-analysis/scripts/create_model.R 'mice' $i .5 1000000
done
```


```{bash}
# Concatenating all of the results files for each tissue
bash rat-genomic-analysis/scripts/make_all_results.sh 'mice' './Results/all_Results_DO_mice' 0.5 '1KG_snps'
bash rat-genomic-analysis/scripts/make_all_betas.sh 'mice' './Results/all_betas_DO' 0.5 '1KG_snps'
bash rat-genomic-analysis/scripts/make_all_logs.sh 'mice' './Results/all_logs_DO'
bash rat-genomic-analysis/scripts/make_all_covariances.sh 'mice' './Results/all_covariances_DO' 0.5 '1KG_snps'
```


```{bash}
# Putting these into sqllite databases
python rat-genomic-analysis/scripts/make_sqlite_db.py --output "mice_DO_output_db.db" --results "./Results/all_Results_DO_mice" --construction "./Results/all_logs_DO" --betas "./Results/all_betas_DO" --meta "./Results/allMetaData/mice_DO.allMetaData.txt"
```


```{r}
#compare generated R2 with what Dan generated
pp_mice <- read_tsv(data.dir %&% "all_Results_mice", col_names = TRUE)
dan_mice <- read_tsv(data.dir %&% "mice_predicatibility.txt", col_names = TRUE) 
genome_h2 <- read_tsv(data.dir %&% "mice_pancreas_h2.txt", col_names = TRUE)



all_mice <- inner_join(pp_mice, dan_mice, by = "gene")
cor.test(all_mice$R2, all_mice$pred.perf.R2)

h2_r2 <- inner_join(genome_h2, pp_mice, by = "gene")
cor.test(h2_r2$h2, h2_r2$R2, method = "spearman")

#plot predicted R2
ggplot(all_mice, aes(R2, pred.perf.R2)) + geom_point() + geom_abline() + xlab("R2 generated by Prediction Pipeline") + ylab("R2 from Dan's Models")

#Plot Genomewide H2 vs predicted R2
ggplot(h2_r2, aes(R2, h2)) + geom_point() + geom_abline() + xlab("R2 generated by Prediction Pipeline") + ylab("Genomewide H2 from GCTA")
```


Running PrediXcan
```{r}
# Writing genotype files for each expression file to be used in PrediXcan args
snpMap <- phyMap %>% select(c(snp, varID, refAllele, effectAllele))
snpMap <- cbind(MAF = 0, snpMap)
colnames(snpMap)[2] = "marker"
geno_mice <- inner_join(snpMap, geno, by = "marker")
geno_mice <- geno_mice %>% select(c(chr, varID, pos, refAllele, effectAllele, MAF, colnames(geno_mice)[8:ncol(geno_mice)]))
write_tsv(geno_mice, "./genotypes/genotype_DO_predixcan.txt", col_names = FALSE)

#run split command again to get genotype by chr
```


```{r}
#Write models so that column names match what prediXcan is looking for
filename <- "mice_DO_output_db.db"
sqlite.driver <- dbDriver("SQLite")
conn <- dbConnect(RSQLite::SQLite(), filename)

extra <- dbGetQuery(conn, 'select * from extra')
weights <- dbGetQuery(conn, 'select * from weights')
dbDisconnect(conn)
colnames(extra) = c("gene", "genename", "pred.perf.R2", "n.snps.in.model", "pred.perf.pval")
extra <- cbind(extra, pred.perf.qval = NA)

model_db = MODEL %&% "/mice_DO_predixcan_db.db"
conn <- dbConnect(RSQLite::SQLite(), model_db)
dbWriteTable(conn, "weights", weights)
dbWriteTable(conn, "extra", extra)
```


```{bash}
#run prediXcan
conda activate /gpfs/data/im-lab/nas40t2/natasha/envs/predixcan/
export METAXCAN=/gpfs/data/im-lab/nas40t2/natasha/GTEX_Analysis/MetaXcan/software
export GENO=/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/prediction_pipeline/DOB_pipeline/genotypes
export MODEL=/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/prediction_pipeline/DOB_pipeline

python $METAXCAN/Predict.py \
--model_db_path $MODEL/mice_DO_predixcan_db.db \
--text_genotypes  \
 $GENO/geno_predixcan.chr1.txt \
 $GENO/geno_predixcan.chr2.txt \
 $GENO/geno_predixcan.chr3.txt \
 $GENO/geno_predixcan.chr4.txt \
 $GENO/geno_predixcan.chr5.txt \
 $GENO/geno_predixcan.chr6.txt \
 $GENO/geno_predixcan.chr7.txt \
 $GENO/geno_predixcan.chr8.txt \
 $GENO/geno_predixcan.chr9.txt \
 $GENO/geno_predixcan.chr10.txt \
 $GENO/geno_predixcan.chr11.txt \
 $GENO/geno_predixcan.chr12.txt \
 $GENO/geno_predixcan.chr13.txt \
 $GENO/geno_predixcan.chr14.txt \
 $GENO/geno_predixcan.chr15.txt \
 $GENO/geno_predixcan.chr16.txt \
 l \
 $GENO/geno_predixcan.chr18.txt \
 $GENO/geno_predixcan.chr19.txt \
 $GENO/geno_predixcan.chr20.txt \
--text_sample_ids $GENO/samples_DO_file \
--prediction_output mice_DO__predict.txt  \
--prediction_summary_output mice_DO__summary.txt \
--throw
```


```{r}
#match phenotype file to ids in predixcan results
predict_res <- read_tsv("mice_DO__predict.txt", col_names = TRUE)
pheno <- read_csv(box_dir %&% "Attie-232_Attie_DO_Islets-phenotypes.csv", col_names = TRUE)
pheno <- pheno %>% filter(mouse %in% predict_res$FID )

write_tsv(pheno, box_dir %&% "Attie-232-phenotypes.predixcan.txt", col_names = TRUE)
```


```{r}
#run asssociation in prediXcan
 RESULTS = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/prediction_pipeline/DOB_pipeline/assoc_files/"
 METAXCAN = "/gpfs/data/im-lab/nas40t2/natasha/GTEX_Analysis/MetaXcan/software/"
 EXPR = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/prediction_pipeline/DOB_pipeline/"
pheno <- read_tsv(box_dir %&% "Attie-232-phenotypes.predixcan.txt", col_names = TRUE)

for(i in 2:length(colnames(pheno))){
trait <- colnames(pheno)[i]
runLOC <- "python3 " %&% METAXCAN %&% "PrediXcanAssociation.py " %&% "--expression_file " %&% EXPR %&% "mice_DO__predict.txt --input_phenos_file " %&% box_dir %&% "Attie-232-phenotypes.predixcan.col.txt " %&% "--input_phenos_column " %&% i %&%  " --output " %&% RESULTS %&% "mice_DO__association_" %&% trait %&% "_.txt --verbosity 9 --throw"
system(runLOC)
}
```

