---
title: "rat_prediction_pipeline_nested_R2"
author: "Natasha Santhanam"
date: "9/28/2021"
output: html_document
---

```{r}
suppressMessages(library(dplyr))
suppressMessages(library(glmnet))
suppressMessages((library(reshape2)))
suppressMessages(library(methods))
suppressMessages(library(RSQLite))
"%&%" <- function(a,b) paste(a,b, sep = "")

dir <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/"
source("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/mice_expression/prediction_pipeline/rat-genomic-analysis/PalmerLab_Rat_Analysis.Rmd")
```

```{r generate covariate file, eval=FALSE}
peer_factorsAc <- read.csv(file = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/expression/peerAc/X.csv", header = FALSE)
colnames(peer_factorsAc) <- c('PF1', 'PF2', 'PF3', 'PF4', 'PF5', 'PF6', 'PF7')
rownames(peer_factorsAc) <- rownames(gexAc_transpose)
covariatesAc <- cbind(covariatesAc, peer_factorsAc)
rownames(covariatesAc) <- covariatesAc$rfid

covariatesAc <- covariatesAc %>% select(-c(rfid))

#write.table(covariatesAc, "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/expression/covariatesAc.txt", row.names=TRUE, col.names=TRUE, quote=FALSE)
```



File to run nested cv elastic net for one tissue and chr at a time
```{r gtex_tiss_chrom_training.R, eval=FALSE}
source("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/PredictDB-Tutorial/code/gtex_v7_nested_cv_elnet.R")
"%&%" <- function(a,b) paste(a,b, sep='')

argv <- commandArgs(trailingOnly = TRUE)
chrom <- argv[1]

tiss <- argv[2]


snp_annot_file <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/genotype_files/snp_annot_by_chr/snp_annot.chr" %&% chrom %&% ".RDS"
gene_annot_file <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/genotype_files/snp_annot_by_chr/gene_annotation.RDS"
genotype_file <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/genotype_files/rat_genos_by_chr/genotype.chr" %&% chrom %&% ".txt"

expression_file <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/expression/gex" %&% tiss  %&% ".txt"
covariates_file <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/expression/covariates" %&% tiss %&% ".txt"
prefix <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/nested_R2_rats/" %&% tiss %&% "/Model_training"

main(snp_annot_file, gene_annot_file, genotype_file, expression_file, covariates_file, as.numeric(chrom), prefix, null_testing=FALSE)
```



```{bash, eval=FALSE}
for i in {1..20}
do
  Rscript /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/PredictDB-Tutorial/code/gtex_tiss_chrom_training.R $i "Ac"
done
```


```{r make sql database, eval=FALSE}
driver <- dbDriver('SQLite')
prefix <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/nested_R2_rats/"
model_summaries <- read_tsv(prefix %&% 'Il/Model_training_chr1_model_summaries.txt',col_names = TRUE)
tiss_summary <- read_tsv(prefix %&% 'Il/Model_training_chr1_summary.txt', col_names = TRUE)
  
n_samples <- tiss_summary$n_samples
  
for (i in 2:20) {
  model_summaries <- rbind(model_summaries,
                            read_tsv(prefix %&% 'Il/Model_training_chr'%&%as.character(i) %&% '_model_summaries.txt', col_names = TRUE))
  tiss_summary <- rbind(tiss_summary,
                             read_tsv(prefix %&% 'Il/Model_training_chr' %&% as.character(i) %&% '_summary.txt', col_names = TRUE))
  
}
  
model_summaries <- rename(model_summaries, gene = gene_id)

# Create a database connection
conn <- dbConnect(drv = driver, '/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/models/Il_nested_db.d')
dbWriteTable(conn, 'model_summaries', model_summaries, overwrite = TRUE)
dbExecute(conn, "CREATE INDEX gene_model_summary ON model_summaries (gene)")

# Weights Table -----
weights <- read_tsv(prefix %&% 'Il/Model_training_chr1_weights.txt', col_names = TRUE)
for (i in 2:20) {
  weights <- rbind(weights,
              read_tsv(prefix %&% 'Il/Model_training_chr' %&% as.character(i) %&% '_weights.txt', col_names = TRUE))
  
}
  
weights <- rename(weights, gene = gene_id)
dbWriteTable(conn, 'weights', weights, overwrite = TRUE)
dbExecute(conn, "CREATE INDEX weights_rsid ON weights (rsid)")
dbExecute(conn, "CREATE INDEX weights_gene ON weights (gene)")
dbExecute(conn, "CREATE INDEX weights_rsid_gene ON weights (rsid, gene)")

# Sample_info Table ----
sample_info <- data.frame(n_samples = n_samples, population = 'rats') # Provide the population info
dbWriteTable(conn, 'sample_info', sample_info, overwrite = TRUE)
  
# Construction Table ----
construction <- tiss_summary %>%
                    select(chrom, cv_seed) %>%
                    rename(chromosome = chrom)

dbWriteTable(conn, 'construction', construction, overwrite = TRUE)
dbDisconnect(conn)
```


# Compare nested models to those generated by elasticnet 
```{r data wrangling, eval=FALSE }
#filter for genes that are not NA
model_nested <- model_nested %>% filter(!is.na(cv_R2_avg))
Ac_h2 <- read_tsv("/Users/natashasanthanam/Github/rat-genomic-analysis/data/Ac_h2_annot.txt", col_names = TRUE)

#read in R2 for non nested
  filename <- "/Users/natashasanthanam/Box/imlab-data/data-Github/rat-genomic-analysis/sql/Il_output_db.db"
sqlite.driver <- dbDriver("SQLite")
  conn <- dbConnect(RSQLite::SQLite(), filename)
  extra <- dbGetQuery(conn, 'select * from extra')
dbDisconnect(conn)  
```

```{r compare R2 between the two}
compare_both <- inner_join(model_nested, extra, by = "gene")
cor.test(compare_both$R2, compare_both$cv_R2_avg)
```

```{r}
ggplot(compare_both, aes(cv_R2_avg, R2)) + geom_point()
```
```{r ordered heritability plot}
load_herit <- function(df){
  df$CI_width <- df$se * 1.96
  df <- df[order(df$h2),]
  df$index <- 1:nrow(df)
  return(df)
}

A_df_Ac <- load_herit(Ac_h2) %>% rename(gene = ensembl_gene_id)
A_df_Ac <- inner_join(A_df_Ac, model_nested, by = "gene")
plt_1 <- (ggplot(data = A_df_Ac, aes(x = index))
          + geom_ribbon(aes(ymax = h2 + CI_width, ymin = h2 - CI_width),
                         alpha = 0.25)
          + geom_point(aes(x=index, y=cv_R2_avg), colour = "red", size = 0.2)
          + geom_line(aes(y = h2))
          + geom_hline(yintercept = 0, linetype=2)
          + labs(x = 'Genes Sorted by Heritability',
                 y = 'Heritability',
                 title = "Ac")
          + ylim(-0.5,1)
          + annotate("text", x = 1200, y = 0.9, label = "Mean h2 =  0.08943895", size = 2)
          + annotate("text", x = 1200, y = 0.8, label = "Mean r2 =   -0.03691824", size = 2))
plt_1

A_df_Ac <- load_herit(Ac_h2) %>% rename(gene = ensembl_gene_id)
A_df_Ac <- inner_join(A_df_Ac, extra, by = "gene")

plt_2 <- (ggplot(data = A_df_Ac, aes(x = index))
          + geom_ribbon(aes(ymax = h2 + CI_width, ymin = h2 - CI_width),
                         alpha = 0.25)
          + geom_point(aes(x=index, y=R2), colour = "red", size = 0.2)
          + geom_line(aes(y = h2))
          + geom_hline(yintercept = 0, linetype=2)
          + labs(x = 'Genes Sorted by Heritability',
                 y = 'Heritability',
                 title = "Ac")
          + ylim(-0.5,1)
          + annotate("text", x = 1200, y = 0.9, label = "Mean h2 =  0.08943895", size = 2)
          + annotate("text", x = 1200, y = 0.8, label = "Mean r2 =  0.08507938", size = 2))
plt_2

```

