---
title: "sabrina-analysis"
author: "sabrina-mi"
date: "2020-08-07"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

```{r}
library(tidyverse)
library(devtools)
library(broom)
library(data.table)
library(RSQLite)
```

Data from [here](https://uchicago.app.box.com/folder/102043737114) - genoGex.Rdata has everything we need in it
There are 5 'gex' RDS files which are the gene expressions for the 5 different tissues, the 'gtf' is the gene annotation, 'phyMap' is the snp annotation, and 'geno' is the genotype matrix


# Create Dosage file

```{r, eval=FALSE}
load("/Users/sabrinami/Github/rat-genomic-analysis/data/Data-From-Abe-Palmer-Lab/Rdata/genoGex.RData")
```

```{r, eval=FALSE}
# Making the snp annotation in the correct format for the pipeline

phyMap <- within(phyMap,  varID <- paste(Chr, Pos, Ref, Alt, sep="_"))
rownames(geno) = paste(phyMap$Chr, phyMap$Pos, phyMap$Ref, phyMap$Alt, sep="_")
```

```{r, eval=FALSE}
library(readxl)

pheno <- read_excel("/Users/sabrinami/Github/rat-genomic-analysis/data/Data-From-Abe-Palmer-Lab/Final_P50_traits/P50_raw_traits.xlsx")
dim(pheno)
pheno_values <- apply(pheno, 2, function(x) sum(!is.na(x)))
hist(pheno_values, breaks=20)


samples <- data.frame("FID" = crf$rfid, "IID" = crf$rfid)

```



```{r, eval=FALSE}
samples_geno <- data.frame(geno[, match(crf$rfid, colnames(geno))])

samples_geno <- mutate(samples_geno, variant_id = row.names(samples_geno), variant_id_dup = variant_id, MAF = 0)
samples_geno <- separate(samples_geno, variant_id_dup, c("chromosome", "position", "allele1", "allele2"), "_")
samples_geno <- select(samples_geno, chromosome, variant_id, position, allele1, allele2, MAF, everything())
write.table(samples_geno, "/Users/sabrinami/Github/rat-genomic-analysis/output/genotypes/chr1_20.txt", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
write.table(samples, "/Users/sabrinami/Github/rat-genomic-analysis/output/genotypes/samples.txt", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)

write.table(crf, "/Users/sabrinami/Github/rat-genomic-analysis/output/phenotypes/crf_MI_active_responses.txt", sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

```

# Run Association

```{bash, eval=FALSE}
python3 /Users/sabrinami/Github/MetaXcan/software/Predict.py \
--model_db_path /Users/sabrinami/Github/rat-genomic-analysis/output/models/Ac.db  \
--text_genotypes /Users/sabrinami/Github/rat-genomic-analysis/output/genotypes/chr1_20.txt.gz \
--text_sample_ids /Users/sabrinami/Github/rat-genomic-analysis/output/genotypes/samples.txt \
--prediction_output /Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Ac__predict.txt \
--prediction_summary_output /Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Ac__summary.txt \
--verbosity 1 \
--throw

```

```{bash, eval=FALSE}
python3 /Users/sabrinami/Github/MetaXcan/software/PrediXcanAssociation.py \
--expression_file /Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Ac__predict.txt \
--input_phenos_file /Users/sabrinami/Github/rat-genomic-analysis/output/phenotypes/crf_MI_active_responses.txt \
--input_phenos_column crf_MI_active_responses \
--output /Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Ac__association.txt \
--verbosity 1 \
--throw

```

## Run MulTiXcan.py

```{bash, eval=FALSE}
python /Users/sabrinami/Github/MetaXcan/software/MulTiXcan.py \
--expression_folder /Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan \
--expression_pattern "(.*)__predict.txt" \
--input_phenos_file /Users/sabrinami/Github/rat-genomic-analysis/output/phenotypes/crf_MI_active_responses.txt \
--input_phenos_column crf_MI_active_responses \
--output /Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/mt_predixcan.txt \
--verbosity 1 \
--throw

```

## Plot z-scores

```{r}
library(data.table)
Ac_association <- fread("/Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Ac__association.txt")
Il_association <- fread("/Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Il__association.txt")
Lh_association <- fread("/Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Lh__association.txt")
Pl_association <- fread("/Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Pl__association.txt")
Vo_association <- fread("/Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Vo__association.txt")
multi_tissue_association <- fread("/Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/mt_predixcan.txt")

```

```{r}
suppressPackageStartupMessages(library(UpSetR))
suppressPackageStartupMessages(library(tidyverse))

scatter_base_theme_ = function(base_size=15) {
  theme_bw(base_size = base_size) +
    theme(panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(colour = "#170a45", size = .5),
          axis.ticks = element_line(colour = "#170a45", size = .2),
          axis.text = element_text(color = '#170a45'))
}

gg_qqplot <- function(ps, ci = 0.95, max_yval = 30) {
  # Many thanks to github user slowkow for this function
  # https://gist.github.com/slowkow/9041570
  ps <- ps[!is.na(ps)]
  n  <- length(ps)
  bf <- (1 - ci) / n
  df <- data.frame(
    observed = -log10(sort(ps)),
    expected = -log10(ppoints(n)),
    clower   = -log10(qbeta(p = (1 - ci) / 2, shape1 = 1:n, shape2 = n:1)),
    cupper   = -log10(qbeta(p = (1 + ci) / 2, shape1 = 1:n, shape2 = n:1))
  )
  df$observed[df$observed > max_yval] <- max_yval
  log10Pe <- expression(paste("Expected -log"[10], plain(P)))
  log10Po <- expression(paste("Observed -log"[10], plain(P)))
  ggplot(df) +
    geom_point(aes(expected, observed), size = 3) +
    geom_abline(intercept = 0, slope = 1, alpha = 0.5) +
    geom_line(aes(expected, cupper), linetype = 2) +
    geom_line(aes(expected, clower), linetype = 2) +
    xlab(log10Pe) +
    ylab(log10Po) +
    geom_hline(yintercept = -log10(bf)) +
    scatter_base_theme_()
}
```

MulTiXcan pvalues:

```{r}
gg_qqplot(multi_tissue_association$pvalue)


```

PrediXcan with Ac tissue model:

```{r}
gg_qqplot(Ac_association$pvalue)

```

PrediXcan with Il tissue model:

```{r}
gg_qqplot(Il_association$pvalue)

```

PrediXcan with Lh tissue model:

```{r}
gg_qqplot(Lh_association$pvalue)

```

PrediXcan with Pl tissue model:

```{r}
gg_qqplot(Pl_association$pvalue)

```

PrediXcan with Vo tissue model:

```{r}
gg_qqplot(Vo_association$pvalue)

```

```{r}
significant_genes_Ac <- Ac_association %>% filter(pvalue < 0.05/nrow(Ac_association)) %>% arrange(pvalue)
significant_genes_Il <- Il_association %>% filter(pvalue < 0.05/nrow(Il_association)) %>% arrange(pvalue)
significant_genes_Lh <- Lh_association %>% filter(pvalue < 0.05/nrow(Lh_association)) %>% arrange(pvalue)
significant_genes_Pl <- Pl_association %>% filter(pvalue < 0.05/nrow(Pl_association)) %>% arrange(pvalue)
significant_genes_Vo <- Vo_association %>% filter(pvalue < 0.05/nrow(Vo_association)) %>% arrange(pvalue)
significant_genes_multi_tissue <- multi_tissue_association %>% filter(pvalue < 0.05/nrow(multi_tissue_association)) %>% arrange(pvalue)

significant_genes <- list(MulTi = significant_genes_multi_tissue$gene,
              Ac = significant_genes_Ac$gene,
              Il = significant_genes_Il$gene,
              Lh = significant_genes_Lh$gene,
              Pl = significant_genes_Pl$gene,
              Vo = significant_genes_Vo$gene)
upset(fromList(significant_genes), order.by = 'freq', empty.intersections = 'on')
```

# Haky's analysis
```{r}

tempo = pheno %>% select(rfid,sex,color,batchnumber) %>% inner_join(Ac_predict,by=c("rfid"="IID"))

## here we are sampling 10 just to get a sense of whether there are any genes associated with color
for(cc in sample(6:8041, 1)){ 
  print(names(tempo)[cc])
  fit = summary(lm(tempo[[cc]]~tempo$color))
  print(fit)
  }

coefarray = array(NA,c(8041-6+1, 4, 4))
dimnames(coefarray)[[1]] = names(tempo)[6:8041]
dimnames(coefarray)[[2]] = rownames(coef(fit))[-1]
dimnames(coefarray)[[3]] = colnames(coef(fit))

genelist = dimnames(coefarray)[[1]]
ngenes = length(genelist)

cont = 1
#for(cc in 6:8041){ 
for(cc in 6:8041){ 
  fit = summary(lm(tempo[[cc]]~tempo$color))
  coefarray[cont, ,]=coef(fit)[-1,]
  cont = cont + 1
  if(cont/100 == round(cont/100)) print(names(tempo)[cc])
}

pthres = 0.05/ngenes ## bonferroni correction
genelist[coefarray[,"tempo$colorblack","Pr(>|t|)"] < pthres]


```

Exploratory analysis of effects of batch, sex, center on phenotypes 
```{r}

tempo2 = pheno  %>% inner_join(Ac_predict,by=c("rfid"="IID"))
npheno = ncol(pheno) - 5
phenomearray = array(NA,c(8041-6+1, npheno, 4))
phenolist = names(pheno)[-c(1:5)]
pp = phenolist[1]
gg = genelist[1]
## does batch have an effect on phenotypes

## this function calculates the p-value fiven the fstatistc
 mypf = function(fit0){
  fstat= fit0$fstatistic
  1 - pf(fstat[1], fstat[2], fstat[3])
}

exclusionlist = c("open_field_totaldistance","open_field_dist2ctrmean","open_field_dist2ctrtotal")
for(pp in phenolist){
  if(!pp %in% exclusionlist){
fit0 = summary( lm(tempo2[[pp]] ~ batchnumber ,data=tempo2 ) )
pval = mypf(fit0)
if(pval<0.001) print(pval)
  }}
 

```


```{r}


##for(pp in phenolist)
##for(gg in genolist)
fit = summary( lm(tempo2[[pp]] ~ batchnumber + sex  + tempo2[[gg]],data=tempo2 ) )

```



