---
title: "sabrina-analysis"
author: "sabrina-mi"
date: "2020-08-07"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

Data from [here](https://uchicago.app.box.com/folder/102043737114) - genoGex.Rdata has everything we need in it
There are 5 'gex' RDS files which are the gene expressions for the 5 different tissues, the 'gtf' is the gene annotation, 'phyMap' is the snp annotation, and 'geno' is the genotype matrix


# Create Dosage file

```{r, eval=FALSE}
load("/Users/sabrinami/Github/rat-genomic-analysis/data/Data-From-Abe-Palmer-Lab/Rdata/genoGex.RData")
```

```{r, eval=FALSE}
# Making the snp annotation in the correct format for the pipeline

phyMap <- within(phyMap,  varID <- paste(Chr, Pos, Ref, Alt, sep="_"))
rownames(geno) = paste(phyMap$Chr, phyMap$Pos, phyMap$Ref, phyMap$Alt, sep="_")
```

```{r}
library(readxl)

pheno <- read_excel("/Users/sabrinami/Github/rat-genomic-analysis/data/Data-From-Abe-Palmer-Lab/Final_P50_traits/P50_raw_traits.xlsx")
dim(pheno)
pheno_values <- apply(pheno, 2, function(x) sum(!is.na(x)))
hist(pheno_values, breaks=20)


```



```{r, eval=FALSE}
samples <- data.frame("FID" = pheno$rfid, "IID" = pheno$rfid)

samples_geno <- geno[, match(colnames(geno), pheno$rfid)]
samples_geno <- cbind(samples_geno, row.names(geno), row.names(geno), 0)
samples_geno <- separate(samples_geno, variant_id_dup, c("chromosome", "position", "allele1", "allele2"), "_")
samples_geno <- select(samples_geno, chromosome, variant_id, position, allele1, allele2, MAF, everything())

write.table(samples_geno, "/Users/sabrinami/Github/rat-genomic-analysis/output/genotypes/chr1_20.txt", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
write.table(samples, "/Users/sabrinami/Github/rat-genomic-analysis/output/genotypes/samples.txt", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)

```

# Reformat Prediction Model
```{r, eval=FALSE}
library(RSQLite)
library(tidyverse)
conn <- dbConnect(RSQLite::SQLite(), "/Users/sabrinami/Github/rat-genomic-analysis/data/sql/Ac_output_db.db")
weights <- dbGetQuery(conn, 'select * from weights')
weights <- weights %>% mutate(varID = rsid) %>% select(gene, rsid, varID, ref_allele, eff_allele, weight)

extra <- dbGetQuery(conn, 'select * from extra')
extra <- extra %>% rename(n.snps.in.model = n.snps, pred.perf.pval = pval, pred.perf.R2 = R2) %>% mutate(pred.perf.qval = NA)
extra <- extra[, c(1,2,4,3,5,6)]
dbWriteTable(conn, 'weights', weights, overwrite=TRUE)
dbWriteTable(conn, 'extra', extra, overwrite=TRUE)
dbDisconnect(conn)

```

# Run Predict

```{bash, eval=FALSE}
conda activate imlabtools

python3 /Users/sabrinami/Github/MetaXcan/software/Predict.py \
--model_db_path /Users/sabrinami/Github/rat-genomic-analysis/data/sql/Ac_output_db.db  \
--text_genotypes /Users/sabrinami/Github/rat-genomic-analysis/output/genotypes/chr1_20.txt.gz \
--text_sample_ids /Users/sabrinami/Github/rat-genomic-analysis/output/genotypes/samples.txt \
--prediction_output /Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Ac__predict.txt \
--prediction_summary_output /Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Ac__summary.txt \
--verbosity 1 \
--throw

```


# Haky's analysis
```{r}
library(tidyverse)
library(data.table)
Ac_predict <- fread("/Users/sabrinami/Github/rat-genomic-analysis/Results/PrediXcan/Ac__predict.txt")

tempo = pheno %>% select(rfid,sex,color,batchnumber) %>% inner_join(Ac_predict,by=c("rfid"="IID"))

## here we are sampling 10 just to get a sense of whether there are any genes associated with color
for(cc in sample(6:8041, 1)){ 
  print(names(tempo)[cc])
  fit = summary(lm(tempo[[cc]]~tempo$color))
  print(fit)
  }

coefarray = array(NA,c(8041-6+1, 4, 4))
dimnames(coefarray)[[1]] = names(tempo)[6:8041]
dimnames(coefarray)[[2]] = rownames(coef(fit))[-1]
dimnames(coefarray)[[3]] = colnames(coef(fit))

genelist = dimnames(coefarray)[[1]]
ngenes = length(genelist)

cont = 1
#for(cc in 6:8041){ 
for(cc in 6:8041){ 
  fit = summary(lm(tempo[[cc]]~tempo$color))
  coefarray[cont, ,]=coef(fit)[-1,]
  cont = cont + 1
  if(cont/100 == round(cont/100)) print(names(tempo)[cc])
}

pthres = 0.05/ngenes ## bonferroni correction
colorblack_genes <- genelist[coefarray[,"tempo$colorblack","Pr(>|t|)"] < pthres]


```

Exploratory analysis of effects of batch, sex, center on phenotypes 
```{r}

tempo2 = pheno  %>% inner_join(Ac_predict,by=c("rfid"="IID"))
npheno = ncol(pheno) - 5
phenomearray = array(NA,c(8041-6+1, npheno, 4))
phenolist = names(pheno)[-c(1:5)]
pp = phenolist[1]
gg = genelist[1]
## does batch have an effect on phenotypes

```

```{r, eval=FALSE}
## this function calculates the p-value fiven the fstatistc
 mypf = function(fit0){
  fstat= fit0$fstatistic
  1 - pf(fstat[1], fstat[2], fstat[3])
}

exclusionlist = c("open_field_totaldistance","open_field_dist2ctrmean","open_field_dist2ctrtotal")
for(pp in phenolist){
  if(!pp %in% exclusionlist){
fit0 = summary( lm(tempo2[[pp]] ~ batchnumber ,data=tempo2 ) )
pval = mypf(fit0)
if(pval<0.001) print(pval)
  }}
 

```

# Phenotype Gene Association

```{r, eval=FALSE}

phenomearray = array(NA,c(8041-6+1, npheno, 4))
dimnames(phenomearray)[[1]] = genelist
dimnames(phenomearray)[[2]] = phenolist
dimnames(phenomearray)[[3]] = colnames(coef(fit))
pcount = 1
gcount = 1
for(pp in phenolist) {
  print(pp)
  gcount = 1
  for(gg in genelist) {
    fit = summary( lm (tempo2[[pp]] ~ tempo2[[gg]], data=tempo2))
    phenomearray[gcount, pcount, ] = coef(fit)[-1,]
    gcount = gcount + 1
  }
  pcount = pcount + 1
}

saveRDS(phenomearray, "/Users/sabrinami/Github/rat-genomic-analysis/output/phenomearray.rds")


```

# Annotate Genes
```{r}
library(biomaRt)
ensembl = useMart(biomart="ENSEMBL_MART_ENSEMBL", dataset = "rnorvegicus_gene_ensembl")
orth.rat = getBM( attributes=
                    c("ensembl_gene_id",
                      "chromosome_name",
                      "start_position",
                      "end_position"),
                  values =TRUE,
                  mart = ensembl,
                  bmHeader=FALSE)

gene_annot = data.frame(gene = genelist)
gene_annot = inner_join(gene_annot, orth.rat, by = c("gene" = "ensembl_gene_id"))
```

# Plots

```{r}
library(qqman)
phenomearray2 <- readRDS("/Users/sabrinami/Github/rat-genomic-analysis/output/phenomearray.rds")
pcount = 1
for(pcount in 1:npheno) {
  
  
  pvalues <- data.frame(pval = phenomearray2[,pcount,4])
  pvalues$gene <- rownames(pvalues)
  pvalues <- inner_join(pvalues, gene_annot, by = "gene")
  pvalues$chromosome_name <- as.integer(pvalues$chromosome_name)
  pvalues$start_position <- as.integer(pvalues$start_position)
  pvalues$end_position <- as.integer(pvalues$end_position)
  
  PHENO <- dimnames(phenomearray2)[[2]][pcount]
  MAN <- glue::glue("/Users/sabrinami/Github/rat-genomic-analysis/Results/plots/{PHENO}_manhattan.png")
  QQ <- glue::glue("/Users/sabrinami/Github/rat-genomic-analysis/Results/plots/{PHENO}_qq.png")
  png(MAN)
  manhattan(pvalues, chr = "chromosome_name", bp = "start_position", p = "pval", snp = "gene")
  dev.off()
  
  
  png(QQ)
  qq(pvalues$pval)
  dev.off()
}


```

